python/ray/data/_internal/batcher.py

```diff
@@ -222,9 +230,51 @@ def add(self, block: Block):
         Args:
             block: Block to add to the shuffle buffer.
         """
-        if BlockAccessor.for_block(block).num_rows() > 0:
+        # Because Arrow tables are memory mapped, blocks in the local shuffle buffer
+        # resides in object store memory and not local heap memory. So, if you specify a
+        # large buffer size and there isn't enough object store memory on the node, you
+        # encounter spilling.
+        if (
+            self._estimated_max_buffer_nbytes is not None
+            and self._estimated_max_buffer_nbytes > self._total_object_store_nbytes
```

Comment:
nit, 1) this if statement can be put under `if BlockAccessor.for_block(block).num_rows() > 0:` and after `after_block`. because this will guarantee that _estimated_max_buffer_nbytes is not None
2) we can skip calculating `_estimated_max_buffer_nbytes` after it's calculated once.